File: DataReconcile_Python/estAccuracyIncreaseByDR.py
import numpy as np

def get_imag_numeric_derivative(func, params, params_, deriv_param_num, equation_num):
    alpha = np.zeros_like(params, dtype = float)
    
    alpha[deriv_param_num] = params[deriv_param_num] * 1e-3 + 1e-10
    #df = [func(params + alpha * 1j, params_) - func(params, params_)]
    df = np.subtract(func(params + alpha * 1j, params_), func(params, params_))
    if alpha[deriv_param_num] == 0:
        df_dx = 0
    else:
        df_dx = np.imag(df[equation_num]) / alpha[deriv_param_num]
    return df_dx

def estAccuracyIncreaseByDR(dep_func, ChosenMode, vals_for_reconc, params_, cov_matrix):
    num_equations = len(dep_func(vals_for_reconc, params_))
    num_vals = len(vals_for_reconc)

    variances_of_DR_result = np.zeros(num_vals)
    accuracy_increase_ratio = np.zeros(num_vals)
    dF_dxj = np.zeros(num_equations)
    Jacobian = np.zeros((num_equations, num_vals - 1))

    for j in range(num_vals):
        # nums_without_j_th = list(range(1, j)) + list(range(j + 1, num_vals))
        nums_without_j_th = list(range(0, j)) + list(range(j + 1, num_vals))
        tmp_mx = cov_matrix.copy()
        tmp_mx = np.delete(tmp_mx, j, axis=0)
        tmp_mx = np.delete(tmp_mx, j, axis=1)
        cov_matrix_without_j_th = tmp_mx
        for i in range(num_equations):
            dF_dxj[i] = get_imag_numeric_derivative(dep_func, vals_for_reconc, params_, j, i)
            for j_ in range(num_vals - 1):
                Jacobian[i][j_] = get_imag_numeric_derivative(dep_func, vals_for_reconc, params_, nums_without_j_th[j_], i)
        if ChosenMode == 'LS':
            linear_est_of_indirect_measurement_variance = np.abs(dF_dxj @ Jacobian) @ cov_matrix_without_j_th @ np.abs(Jacobian.T @ dF_dxj) / ((dF_dxj @ dF_dxj) ** 2)
        elif ChosenMode == 'WLS':
            var_of_delta_x_j = np.zeros(num_equations)
            weights_for_WLS = np.zeros(num_equations)
            for j_ in range(num_equations):
                sum_df_j_dz = np.sum(Jacobian[j_, :] ** 2 * np.diag(cov_matrix_without_j_th))
                var_of_delta_x_j[j_] = dF_dxj[j_] ** (-2) * sum_df_j_dz
            for j_ in range(num_equations):
                tmp_w = var_of_delta_x_j[j_] / var_of_delta_x_j
                tmp_w[np.isnan(tmp_w)] = 1
                weights_for_WLS[j_] = 1 / np.sum(tmp_w)
            linear_est_of_indirect_measurement_variance = np.square(weights_for_WLS) @ var_of_delta_x_j
        else:
            print('Unknown mode for linear estimation of variance of system solution')
        if linear_est_of_indirect_measurement_variance == 0:
            print('The true value of {}-th measurand can be derived from given dependency equations. No reconciliation is needed for its measurement result.'.format(j))
        accuracy_increase_ratio[j] = np.sqrt(1 + cov_matrix[j][j] / linear_est_of_indirect_measurement_variance)
        variances_of_DR_result[j] = cov_matrix[j][j] / accuracy_increase_ratio[j] ** 2

    return accuracy_increase_ratio, variances_of_DR_result

File: DataReconcile_Cpp/estAccuracyIncreaseByDR.cpp
#include <iostream>
#include <vector>
#include <complex>
#include <cmath>
#include <Eigen/Dense>

using namespace std;
using namespace Eigen;

// Function to compute the imaginary part of the numerical derivative
double get_imag_numeric_derivative(const function<VectorXcd(const VectorXd&, const VectorXd&)>& func,
    const VectorXd& params, const VectorXd& params_,
    int deriv_param_num, int equation_num) {
    VectorXd alpha = VectorXd::Zero(params.size());
    alpha(deriv_param_num) = params(deriv_param_num) * 1e-3 + 1e-10;

    VectorXcd df = func(params + alpha * complex<double>(0, 1), params_) - func(params, params_);

    if (alpha(deriv_param_num) == 0) {
        return 0;
    }
    else {
        return imag(df(equation_num)) / alpha(deriv_param_num);
    }
}

// Function to estimate accuracy increase by data reconciliation
pair<VectorXd, VectorXd> estAccuracyIncreaseByDR(const function<VectorXd(const VectorXd&, const VectorXd&)>& dep_func,
    const string& ChosenMode, const VectorXd& vals_for_reconc,
    const VectorXd& params_, const MatrixXd& cov_matrix) {
    int num_equations = dep_func(vals_for_reconc, params_).size();
    int num_vals = vals_for_reconc.size();

    VectorXd variances_of_DR_result = VectorXd::Zero(num_vals);
    VectorXd accuracy_increase_ratio = VectorXd::Zero(num_vals);
    VectorXd dF_dxj = VectorXd::Zero(num_equations);
    MatrixXd Jacobian = MatrixXd::Zero(num_equations, num_vals - 1);

    for (int j = 0; j < num_vals; ++j) {
        vector<int> nums_without_j_th;
        for (int k = 0; k < num_vals; ++k) {
            if (k != j) {
                nums_without_j_th.push_back(k);
            }
        }

        MatrixXd tmp_mx = cov_matrix;
        tmp_mx.block(j, 0, num_vals - 1, num_vals) = tmp_mx.block(j + 1, 0, num_vals - 1, num_vals);
        tmp_mx.block(0, j, num_vals, num_vals - 1) = tmp_mx.block(0, j + 1, num_vals, num_vals - 1);
        tmp_mx.conservativeResize(num_vals - 1, num_vals - 1);
        MatrixXd cov_matrix_without_j_th = tmp_mx;

        for (int i = 0; i < num_equations; ++i) {
            dF_dxj(i) = get_imag_numeric_derivative([&](const VectorXd& x, const VectorXd& p) {
                return dep_func(x, p).cast<complex<double>>();
                }, vals_for_reconc, params_, j, i);

            for (int j_ = 0; j_ < num_vals - 1; ++j_) {
                Jacobian(i, j_) = get_imag_numeric_derivative([&](const VectorXd& x, const VectorXd& p) {
                    return dep_func(x, p).cast<complex<double>>();
                    }, vals_for_reconc, params_, nums_without_j_th[j_], i);
            }
        }

        double linear_est_of_indirect_measurement_variance;
        if (ChosenMode == "LS") {
            linear_est_of_indirect_measurement_variance = abs((dF_dxj.transpose() * Jacobian) *
                cov_matrix_without_j_th *
                (Jacobian.transpose() * dF_dxj) /
                pow(dF_dxj.squaredNorm(), 2);
        }
        else if (ChosenMode == "WLS") {
            VectorXd var_of_delta_x_j = VectorXd::Zero(num_equations);
            VectorXd weights_for_WLS = VectorXd::Zero(num_equations);

            for (int j_ = 0; j_ < num_equations; ++j_) {
                double sum_df_j_dz = (Jacobian.row(j_).array().square() * cov_matrix_without_j_th.diagonal().array()).sum();
                var_of_delta_x_j(j_) = pow(dF_dxj(j_), -2) * sum_df_j_dz;
            }

            for (int j_ = 0; j_ < num_equations; ++j_) {
                VectorXd tmp_w = var_of_delta_x_j.array() / var_of_delta_x_j(j_);
                tmp_w = tmp_w.array().isNaN().select(1, tmp_w);
                weights_for_WLS(j_) = 1.0 / tmp_w.sum();
            }

            linear_est_of_indirect_measurement_variance = weights_for_WLS.array().square().matrix().dot(var_of_delta_x_j);
        }
        else {
            cerr << "Unknown mode for linear estimation of variance of system solution" << endl;
            return make_pair(VectorXd(), VectorXd());
        }

        if (linear_est_of_indirect_measurement_variance == 0) {
            cout << "The true value of " << j << "-th measurand can be derived from given dependency equations. "
                << "No reconciliation is needed for its measurement result." << endl;
        }

        accuracy_increase_ratio(j) = sqrt(1 + cov_matrix(j, j) / linear_est_of_indirect_measurement_variance);
        variances_of_DR_result(j) = cov_matrix(j, j) / pow(accuracy_increase_ratio(j), 2);
    }

    return make_pair(accuracy_increase_ratio, variances_of_DR_result);
}